rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,68 @@
+
+- Multiple applications share the operating system of the same virtual machine. This means it’s not easy to upgrade a single application at a time. Furthermore, the way our [Puppet code](http://github.com/alphagov/govuk-puppet/) is written makes supporting virtual machines on different operating systems at the same time hard to support.
+- GOV.UK currently uses upstart as our init system, but newer versions of Ubuntu switch to systemd. This means GOV.UK will need to do a significant rewrite of our init scripts as part of an upgrade.
+- Our current version of Puppet is 3.8, which is no longer supported (the current version is 6.13). Support for newer operating system versions is likely to be lacking in unsupported versions of Puppet.
+- Other infrastructure as code tools in use (e.g. [Fabric](https://github.com/alphagov/fabric-scripts)) are also currently using very old versions which are very likely to have problems with newer operating systems.
+
+## Proposal
+
+GOV.UK will adopt a container based hosting environment, such as [GOV.UK PaaS](https://www.cloud.service.gov.uk). The choice of hosting environment is out of scope for this RFC.
+
+GOV.UK applications will be built into container images, which will include the operating system. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.
+
+A containerised runtime aligns with the [GOV.UK Infrastructure Architecture Goals for 2021](
+https://docs.google.com/document/d/1ooN7wkYhEGvceGe9Qz_HNZa-GPtrjzK_vA4vfWYVn4c/edit#heading=h.cdrr7rv9t98f) to isolate applications through containers. This will give us better control of how resources are matched to applications.
+
+The TechOps strategy is to use common components where available. GOV.UK has been following the strategy by using hosted versions of software ([ElasticSearch](https://aws.amazon.com/elasticsearch-service/) and [Postgres](https://aws.amazon.com/rds/)) and services ([Notify](https://www.notifications.service.gov.uk)) where available. Using a containerised runtime is a continuation of this policy.
+
+GOV.UK has already containerised its development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.
+
+### Benefits of this approach
+
+The most immediate benefit of this approach is that it will allow us to incrementally upgrade the operating system version. By moving the applications into containers, GOV.UK can enable a much lower risk upgrade path.
+
+Moving to a container based hosting environment will allow us to remove much of our legacy Puppet and Fabric code. This code is hard to maintain, and hard to hire people with the experience required to improve and upgrade. With industry and government overwhelmingly moving towards containerised platforms, it should be much easier to hire people with these skills.
```

Comment:
Could add that it would allow to test configuration locally? - as we currently no way to test puppet without running it in an environment.

Reply:
I think we're kind of alluding to this below with:

> Running containers in production allows for a reduction in the difference between development environments and production

It would be good to call out being able to test configuration changes a bit more explicitly, although in practice there may still be a fair amount of configuration that's different between development and production (depending on which platform we go with).